### ISSUE TITLE
Thoughts on "Detecting" Supported Layers

### ISSUE THREAD

**Issue by Proryanator** [hollance/neural-engine#35]
Hey there! This repo is super amazing, thanks for putting all these findings together.

Was reading this sub-section (link below) on which layers are unsupported, and I had an idea on how to programmatically identify them.

https://github.com/hollance/neural-engine/blob/master/docs/unsupported-layers.md

This part here: "S → U → S → U → S → U" abour swapping supported/unsupported layers made me wonder. In theory, if we take a layer X from a CoreML model we have that we do not yet know can run on the ANE, and we make a model that is built solely of that layer, i.e. X -> X -> X ... and we set the compute unit to be CPU/ANE only, this would encourage the Neural Engine to run this model on 1 compute unit since that would be efficient. So, in theory, if you set the compute unit to be CPU && ANE and you see that the layer runs on the ANE, then you will have identified that this op is ANE compatible? 🧐

I'd like to test this theory but wanted to run this by you. Thinking this could be a way to, given a model, programmatically piecemeal individual layers, build a simple repeated layer model, and produce a chart of whether the layer is CPU/GPU/ANE supported (maybe even with statistics).

Would help with identifying areas where a layer could be swapped out/modified to encourage running the model more efficiently on the ANE.

Any thoughts would be appreciated! 😊

**Comment by hollance** [comment]
That sounds like an interesting approach!

**Comment by Proryanator** [comment]
Thanks! Will give this a shot at some point and share what I find.

**Comment by Proryanator** [comment]
@hollance thanks for writing those CoreML survival guides, an excellent supplement for apple's own documentation. I've started working on this, calling it "anetools" 👊

**Comment by Proryanator** [comment]
Using your section of "Model Surgery" I was able to successfully isolate a single layer in a Neural Network! Cool thing is I could run the performance tab on it too.

Going to work on making this generically possible for all layers in the model (and support more than just NeuralNetwork models) but, progress!

<img width="980" alt="Screenshot 2024-05-25 at 11 17 39 PM" src="https://github.com/hollance/neural-engine/assets/8595173/2d621b50-b5b1-4f03-bbca-6614f9976ced">

**Comment by hollance** [comment]
Honestly, it's been too long since I did anything with Core ML so I don't know off the top of my head.

**Comment by Proryanator** [comment]
@hollance was this feature in XCode around when you were working on CoreML?

I saw the apple discussion/article was from WWDC22' but not sure if it was around back then.

It somehow detects the compute unit per layer, and also tells you whether a layer is compatible with a given compute unit. It sometimes doesn't work but when it does it is pretty nice.

https://developer.apple.com/videos/play/wwdc2022/10027/

**Comment by hollance** [comment]
Unfortunately that feature was not available when I was working with Core ML. Would have been useful at the time. :-)

**Comment by Proryanator** [comment]
Yeah I realized this does _most_ of what I'm trying to research 😆 although it does on occasion fail for ML Program models it seems.
